prov/efa/src/efa_hmem.c

```diff
@@ -322,32 +321,39 @@ int efa_hmem_validate_p2p_opt(struct efa_domain *efa_domain, enum fi_hmem_iface
 
 	if (OFI_UNLIKELY(!info->initialized))
 		return -FI_ENODATA;
-	
-	switch (p2p_opt) {
-
-	/* REQUIRED is valid even if an interface doesn't require P2P */
-	case FI_HMEM_P2P_REQUIRED:
-		return info->p2p_supported_by_device ? FI_SUCCESS : -FI_EOPNOTSUPP;
 
 	/*
-	 * Invalidate ENABLED or PREFERRED if an iface requires P2P, or if an iface
-	 * doesn't support P2P at all
+	 * According to fi_setopt() document:
+	 *
+	 *     ENABLED means a provider may use P2P.
+	 *     PREFERED means a provider should prefer P2P if it is available.
+	 *
+	 * These options does not require that p2p is supported by device,
+	 * nor do they prohibit that p2p is reqruied by implementation. Therefore
+	 * they are always supported.
 	 */
-	case FI_HMEM_P2P_PREFERRED:
-	case FI_HMEM_P2P_ENABLED:
-		return info->p2p_supported_by_device && !info->p2p_required_by_impl
-			? FI_SUCCESS
-			: -FI_EOPNOTSUPP;
+	if (p2p_opt == FI_HMEM_P2P_PREFERRED || p2p_opt == FI_HMEM_P2P_ENABLED) {
+		info->p2p_disabled_by_user = false;
```

Comment:
Aside from removing the conditions `info->p2p_supported_by_device && !info->p2p_required_by_impl` to comply with the Libfabric API spec, this line (and 344) are the only real notable changes. This change is necessary since it updates `info->p2p_disabled_by_user` regardless of `p2p_opt`.

Everything else is just style preference. Naturally I think the `switch`...`case` is cleaner/easier to understand.

```c
switch (p2p_opt) {
case FI_HMEM_P2P_REQUIRED:
    if (!info->p2p_supported_by_device)
        return -FI_EOPNOTSUPP;
    /* Fall-through */
case FI_HMEM_P2P_PREFERRED:
case FI_HMEM_P2P_ENABLED:
    info->p2p_disabled_by_user = false;
    return 0;
case FI_HMEM_P2P_DISABLED:
    if (info->p2p_required_by_impl)
        return -FI_EOPNOTSUPP;
    info->p2p_disabled_by_user = true;
    return 0;
}
return -FI_EINVAL;
```

Reply:
Thanks! Switched back to use switch case. I did not use fall through though.